### ISSUE TITLE
[Document Lists] Implement a generic post-fixer ensuring correct model of ListProperties

### ISSUE THREAD

**Issue by arkflpc** [ckeditor/ckeditor5#11167]
Goal: Make sure that all items in a single list (items at the same level) have the same start, reversed, style.

Most likely, the above will (in the future) be ensured by a combination of:

* (a) DocListProperties and DocList commands cooperating together.
* (b) Listeners for `deleteContent` and `insertContent`.
* (c) A post-fixer that we need anyway to handle e.g. RTC-related issues.

For now, let's focus only on the last part – the post-fixer. We need to implement it anyway.

Any UX, performance or logical issues that can't be solved generically by postfixer or are not easy enough, are out of scope. For instance:
- merging two lists
- fixing list after indentation changes
- pasting lists
- collaboration.

We should implement now at least some basic tests for these scenarios. If the post-fixer won't be able to handle them in a reasonable (from UX perspective) way, we can skip them for now and use them as an input when working on the **(a)** and **(b)**.

**Comment by Reinmar** [comment]
F2F discussion notes...

> Goal: Make sure that all items in a single list (items at the same level) have the same start, reversed, style.

The problem with this is that previously (in old lists) we could have different properties only on the top-level (they'd create two different lists). In the new implementation, we also allow this on every level of a list.

Thus, such a post-fixer doesn't (perhaps) make sense now.

Post-fixers that are a must:

*   Removing properties when they are not applicable to the current block or resetting `listStyle` (to the default value) when changing the list type.
*   Adding attributes to items that had none of them (e.g. bare list item, inserted by an integrator). In this case – inherit attributes from a list to which this list was prepended/inserted/appended (if such a list exists already).

We can treat a post-fixer as a tool to ensure things that the schema couldn't. All other rules need to be implemented in more contextual APIs (e.g. enter key some).

... discussion about possible approaches:

*   general post-fixer (as initially proposed)
*   approaching issues one by one (would require prioritizing them in terms of UX value)
*   porting the old post-fixer (with its potential problems)

@arkflpc will continue his research here to figure out the best approach

**Comment by Reinmar** [comment]
> *   porting the old post-fixer (with its potential problems)

What's the cost of this approach? Potentially, the lowest. 

Risks:

*   It will not allow having nested lists with different attributes.
*   The old post-fixer may have some obvious missing rules when confronted with the new lists' structure.

**Comment by Reinmar** [comment]
We discussed this again and @arkflpc proposed two directions:

*   The initially proposed generic post-fixer unifying lists on all levels.
*   A DDD approach – discovering scenarios as we go and covering them with implementation.

The former allows making a quick jump at once, but we don't know the level of UX after that. It may require fine-tuning in next steps.

The latter will be more precise, but we won't be able to understand the whole picture for quite a long time.

After long discussions we decided to go with the first approach.
